### DISCUSSION TITLE
Proxy still needed with HTTP/2 supported by most browsers?

### DISCUSSION THREAD

**Discussion by aherreraGH** [grpc/grpc-web#1348] [Q&A - General] [answered]
Hello everyone, 

_Fairly new to gRPC world here._ 

General question, I've been reading that just about all major browsers (except IE) now offer > 90% support for HTTP/2. So, I'm curious, is the proxy like EnvoyProxy still required? Some articles I reviewed stated the need for Envoy as HTTP2 is not widely accepted. The articles are 2+ years old. 

If a proxy is still required/needed: 
Suggestions on getting golang server grpc to talk with grpc-web? I have tried EnvoyProxy with no success, perhaps there is a better alternative proxy option out there? Could I use built-in proxy similar to ViteJS / Express / Webpack middleware? Anyone have examples with these?

If a proxy is no longer needed: 
How would I go about communicating with a golang grpc server using grpc-web? I did find there is an NPM package for fetch-h2, http2-client, and others. Wondering if they would do the job as mentioned above, whereas browsers now support http2 mostly - from what I read.

Thank you in advance.

**Comment by sampajano** [comment, marked as answer]
@aherreraGH Hi! Thanks for asking! Will answer below 😃

> General question, I've been reading that just about all major browsers (except IE) now offer > 90% support for HTTP/2. So, I'm curious, is the proxy like EnvoyProxy still required? Some articles I reviewed stated the need for Envoy as HTTP2 is not widely accepted. The articles are 2+ years old.

You're indeed right that HTTP/2 is widely available.

Although, the lack of HTTP/2 availability was *NOT* the reason why grpc-web protocol was created in the first place; rather, it is because the gRPC protocol relied on HTTP/2 framing, which is NOT exposed in browsers APIs, and likely never will be.

As it is mentioned in the [protocol doc](https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-WEB.md#design-goals):

> decouple from HTTP/2 framing which is not, and will never be, directly exposed by browsers

As such, a proxy would still be required in the foreseeable future.

---
> If a proxy is still required/needed:

> Suggestions on getting golang server grpc to talk with grpc-web? I have tried EnvoyProxy with no success, perhaps there is a better alternative proxy option out there? 

There is a good amount of proxy options you could consider. See https://github.com/grpc/grpc-web#proxy-interoperability for the list of proxies with grpc-web support.

> Could I use built-in proxy similar to ViteJS / Express / Webpack middleware? Anyone have examples with these?

I'm not aware of any supported built-in proxies as of today. But my knowledge is limited :)

Hope that helps!

**Reply by aherreraGH** [reply]
Thank you!
